### DISCUSSION TITLE
compile time optimisations

### DISCUSSION THREAD

**Discussion by arturbac** [stephenberry/glaze#1511] [Optimization]
from discussion on issue #1501 
here from example_json

### Write insantations

```cpp glz::write_json<BasicStruct &, std::basic_string<char> &>```
Wall Duration	33,266 ms

![obraz](https://github.com/user-attachments/assets/b94bcdc0-f7b6-404d-8895-38f191e9cadf)
------------
Wall Duration | instantation 
---|-----------
 10,695 ms | ```cpp glz::reflect<BasicStruct>``` 
 6,144 ms |  ```cpp glz::for_each_short_circuit<4UL, (lambda at /home/artur/projects/glaze/include/glaze/core/reflect.hpp:316:39)>``` 
8,386 ms | ```cpp "glz::detail::invoke_table<4UL, (lambda at /home/artur/projects/glaze/include/glaze/json/write.hpp:1571:35)>" ```

### Read instantations
```cpp glz::read_json<BasicStruct, std::basic_string<char> &>```
Wall Duration	66,972 ms

![obraz](https://github.com/user-attachments/assets/21637dfc-be07-4f3b-9081-f2f05fb25275)

------------
Wall Duration | instantation 
---|-----------
3 x 12,912 ms | ```cpp glz::detail::front_bytes_hash_info<unsigned short, 4UL> ```
16,521 ms | ```cpp glz::detail::unique_per_length_info<std::array<std::basic_string_view<char>, 4>> ```
5 x ~5,942 ms | ```cpp glz::detail::decode_index<opts{...}, BasicStruct, 1UL, BasicStruct &, glz::context &, const char *&, const char *&> ```

where functions front_bytes_hash_info, unique_per_length_info are called for glz::detail::make_keys_info<4UL> which takes 45,007 ms of wall time

**Comment by arturbac** [comment]
removing span from compile time instantiations and move to pointer + size sometimes can cut time in half. I had to remove span use in consteval in simple enum for that reason (replaced with ptr, size arguemnts to functions)
![obraz](https://github.com/user-attachments/assets/b6782d5d-bdb5-41d1-a393-f2cee33d6545)

wall 12,912 ms "glz::detail::front_bytes_hash_info<unsigned short, 4UL>"
where construction of span calling constructor and initialization takes
half of that time 
span is build wall 6,479 ms "std::span<unsigned long>"
jsut for wall 2,869 ms processing of std::reverse_iterator<__gnu_cxx::__normal_iterator<unsigned long *, std::span<unsigned long>>>
So half the time is span and its reverse iterator, which can be just probably --ptr or at least reverse iterator from ptr, size arguments

ex:
```cpp
const auto bucket = hash % bsize;
               if (contains(std::span{bucket_index.data(), index}, bucket)) {
                  break;
// uses span but can ptr and size
   constexpr bool contains(auto&& data, auto&& val) noexcept
   {
      const auto n = data.size();
      for (size_t i = 0; i < n; ++i) {
         if (data[i] == val) {
            return true;
         }
      }
      return false;
   }
```

**Comment by stephenberry** [comment]
Wow! Yeah, because these are known types, I can probably also make a pseudo-span that isn't templated. But, I'd first go for a pointer and size like you suggested.

Feel free to make pull requests for these changes, but I'll also try to do some -ftime-trace  optimizations when I'm able.

**Reply by arturbac** [reply]
rules I found for writing fast compile time code are:
- think of it as of old c++98 compiler without optimisations at all running code in DEBUG mode with all constructors, destructors being called with no RVO etc
- returning values by modifying pointers or references at arguments is faster than return by value or larger bojects
- use pure pointers and trivial types

**Reply by stephenberry** [reply]
This makes sense and I've noticed the same, so I'll start trying to take this approach in writing the compile time code. I really appreciate it!

**Reply by arturbac** [reply]
FYI: trying to work a little on this

fixing contains to ptr , size , size
33,266 ms vs 31,305 ms so 7% less for ```glz::write_json<BasicStruct &, std::basic_string<char> &>```

foudn nice one thing
take a look at marked red elements
![obraz](https://github.com/user-attachments/assets/649e5ca8-c4ec-490b-a769-2b806986f49c)
left is ```glz::to_tuple<BasicStruct, 4UL>```
right is ```glz::to_tuple<const BasicStruct &, 4UL>```

which does not make any sense to do twice the "same" instantation just because top level has object and deeper same object passed by reference.

I'll take a look what will happen for ```extern T const &``` instead of extern T
it htat will not work then functions needs be fixed, ill try to do it

**Reply by stephenberry** [reply]
I merged in the `contains` update to just use a pointer and sizes. #1512

**Reply by arturbac** [reply]
Not good I don't like to merge things that I am changing, I will have to merge and remove my code then.

**where to put below span and basic_string_view (for use only at compile time) ?**

with them in unique_per_length_info and 
doing array for most cases i got for
```glz::read_json<BasicStruct, std::basic_string<char> &>```
50,591 ms instead of initial 66,972 ms

IMHO all string_views and spans for compile time code we can upgrade to them

```cpp

      std::vector<string_view> strings_;
      std::array<string_view, 128> stringsa_;
      span<string_view> strings;
      if (N < stringsa_.size()) {
         for (size_t i = 0; i < N; ++i) {
            stringsa_[i] = string_view{input_strings[i].data(), uint32_t(input_strings[i].size())};
         }
         strings.data_ = stringsa_.data();
         strings.size_ = N;
      }
      else {
         for (size_t i = 0; i < N; ++i) {
            strings_.emplace_back(string_view{input_strings[i].data(), uint32_t(input_strings[i].size())});
         }
         strings.data_ = strings_.data();
         strings.size_ = strings_.size();
      }
```

```cpp
   template <typename T, typename Traits = std::char_traits<T>>
   struct basic_string_view
   {
      const T* data_;
      uint32_t size_;

      constexpr const T* begin() const noexcept { return data_; }
      constexpr const T* end() const noexcept { return data_ + size_; }
      constexpr uint32_t size() const noexcept { return size_; }
      constexpr uint32_t length() const noexcept { return size_; }
      constexpr bool empty() const noexcept { return size_ == 0; }
      constexpr const T* data() const noexcept { return data_; }
      constexpr auto&& front(this auto&& self) noexcept { return *self.data_; }
      constexpr auto&& back(this auto&& self) noexcept { return *(self.data_ + (self.size_ -1)); }

      constexpr T operator[](size_t sz) const noexcept { return data_[sz]; }
      constexpr bool operator==(const basic_string_view& other) const noexcept
      {
         if (size_ != other.size_) {
            return false;
         }
         return size_ == 0 || Traits::compare(data_, other.data_, size_) == 0;
      }

      constexpr auto operator<=>(const basic_string_view& other) const noexcept
      {
         const auto comp = Traits::compare(data_, other.data_, (size_ < other.size_) ? size_ : other.size_);
         if (comp != 0) {
            return comp <=> 0;
         }
         return size_ <=> other.size_;
      }
   };
   using string_view = basic_string_view<char>;

   template <typename T>
   struct span
   {
      T* data_;
      size_t size_;

      constexpr auto&& front(this auto&& self) noexcept { return *self.data_; }
      constexpr auto&& back(this auto&& self) noexcept { return *(self.data_ + (self.size_ -1)); }
      constexpr auto begin(this auto&& self) noexcept { return self.data_; }
      constexpr auto end(this auto&& self) noexcept { return self.data_ + self.size_; }
   };
```

**Reply by stephenberry** [reply]
> Not good I don't like to merge things that I am changing, I will have to merge and remove my code then.

No, don't worry about merging in newer code, just make a pull request and I can handle any conflicts.

Also, it's helpful if you can make a pull request sooner than later, even if you aren't finished, just so that I can know that you're developing code and try to avoid conflicting.

**Reply by stephenberry** [reply]
In a lot of cases we need the compile time string_views to propagate to runtime, which is why `std::string_view` is desired. But, I look forward to seeing the changes you are proposing.

**Reply by stephenberry** [reply]
I noticed that using std::vector in compile time code often leads to less template instantiations than using std::array<T, N>, because when algorithms like sort are used there are template instantiations needed for every length N for the array, but the vector doesn't require different instantiations for different sizes.

**Reply by arturbac** [reply]
vector -> when using own string view without constructors and default init, it does not require initialization at all so array is just a declaration. You are right that this is not true for std::string_view which default initializes all elements.

**Reply by arturbac** [reply]
You did not answer where to put code of basic_string_view and span, which include ?

**Reply by arturbac** [reply]
> In a lot of cases we need the compile time string_views to propagate to runtime, which is why `std::string_view` is desired. But, I look forward to seeing the changes you are proposing.

We can add implicit operator std::basic_string_view<T>() to own basic_string_view

**Reply by arturbac** [reply]
Btw nice is that having string_view which does not initializes elements by default at compile time this works as a strong verification If Your algorithm works as expected, because when not and some element is not initialized and used it generates compile time error.

**Comment by stephenberry** [comment]
@arturbac

> vector -> when using own string view without constructors and default init, it does not require initialization at all so array is just a declaration. You are right that this is not true for std::string_view which default initializes all elements.

I was actually talking about template instantiations and not initialization with `std::vector` vs` std::array`.

> You did not answer where to put code of basic_string_view and span, which include ?

I would say make includes: `glaze/util/string_view.hpp` and `glaze/util/span.hpp`

> Btw nice is that having string_view which does not initializes elements by default at compile time this works as a strong verification If Your algorithm works as expected, because when not and some element is not initialized and used it generates compile time error.

Excellent point.

**Reply by arturbac** [reply]
> I was actually talking about template instantiations and not initialization with `std::vector` vs` std::array`.

Yes, but what is template instantiation in time trace .. 
it is stack allocation and construction of object. 
So in vector it is stack alloc + 3 members init, in array it is stack alloc and only if elements have non trivial construction then it is initializing whole array to default values or when You annotate to explicit initialize with {}.

Do You agree ?

**Reply by stephenberry** [reply]
No, template instantiation is not the instantiation of the object in binary (assembly), but rather the generation of a new type by the compiler. It only has to occur once per type and is cached, so `std::vector<int>` only needs to be instantiated once and `std::vector<double>` only needs to be instantiated once no matter how many times they are used within a single compilation unit. This is what takes so much time in the time trace, not the initialization of the object.

See:
https://www.modernescpp.com/index.php/template-instantiation/

**Reply by arturbac** [reply]
O I understand what You mean, You are right.
But here is not much difference in instantiation of array<int,N> and vector<int> extcept that array code is many times simpler.
But empalce_back , allcoations  etc still are many time slower than 
setting not initialized array element.
So I don't get where You see potential to have use of vector cheaper than array with simple aggregate structs

btw as You can see there is 50ms vs 66 ms difference in example

**Reply by stephenberry** [reply]
In a large codebase with structs of various sizes we will have `std::array<string_view, 2>`, `std::array<string_view, 3>`, `std::array<string_view, 5>`, `std::array<string_view, 7>`, etc. All these need separate template instantiations, whereas `std::vector<string_view>` only needs one instantiation. This is compounded if we use `std::ranges::sort`, because it needs to instantiate itself with each differently sized array. So, a single `std::array` vs a `std::vector` may be faster, but we need to consider all the template instantiations over the entire codebase.

**Reply by stephenberry** [reply]
I think the best approach is probably to use C style arrays that are stack allocated.

**Reply by arturbac** [reply]
I understand what You meant with array<,N> but You are wrong with unique_per_length_info
because N is not consteval but const auto N = input_strings.size();

so that why i used this below, btw is this N really need to be any size_t ?
because this is pure consteval and none of Your tests needs more than 128, so maybe we can remove at all vector here ?

```cpp
      std::vector<string_view> strings_;
      std::array<string_view, 128> stringsa_;
      span<string_view> strings;
      if (N < stringsa_.size()) {
         for (size_t i = 0; i < N; ++i) {
            stringsa_[i].data_ = input_strings[i].data();
            stringsa_[i].size_ = uint32_t(input_strings[i].size());
         }
         strings.data_ = stringsa_.data();
         strings.size_ = N;
      }
      else {
         for (size_t i = 0; i < N; ++i) {
            strings_.emplace_back(string_view{input_strings[i].data(), uint32_t(input_strings[i].size())});
         }
         strings.data_ = strings_.data();
         strings.size_ = strings_.size();
      }
```
and this comment is wrong
// This could be a std::array, but each length N for std::array causes unique template instaniations

**Reply by arturbac** [reply]
> I think the best approach is probably to use C style arrays that are stack allocated.

I is hard for me to guess how far You want to push on performance here
when using custom span i proposed it gives clear code but has some penalty when using member functions
on the other hand You can remove almost any begin, end, empty etc and achieve higher performance at a cost of unclear code

**Reply by stephenberry** [reply]
> I understand what You meant with array<,N> but You are wrong with unique_per_length_info
because N is not consteval but const auto N = input_strings.size();

You can pass N in via `template <size_t N>`, which I tested and allowed the use of `std::array`, but then I was seeing more template instantiations due to `std::array` of different sizes.

**Reply by stephenberry** [reply]
I don't think using pointers and sizes is too ugly, especially if you use non-template helper functions. In critical compilation places I'm fine with sacrificing some readability for performance. I'm finishing up a faster to compile `sort` function.

**Reply by stephenberry** [reply]
Also, check out https://www.build-bench.com, which can be really helpful for profiling build times of bits of code.

**Comment by stephenberry** [comment]
@arturbac
I want to make you aware of this active pull request I plan to merge in soon: #1514 
This fixes the duplicate `to_tuple` instantiation that you had brought up. It also moves the `tuple` class out of the `tuplet` namespace, which makes the -ftime-trace results a good bit smaller.

**Reply by arturbac** [reply]
I just think it does not make a sense when I dig out finding issues/potential optimisation places examine how to fix it and then You are faster committing a fix for it.

**Reply by stephenberry** [reply]
I'm trying to fix the trivial things so that you don't have to change as much code. It's better to solve one thing at a time than everything at once.

**Reply by stephenberry** [reply]
If you made pull requests for these fixes then I could contribute and help merge them faster.
